### ISSUE TITLE
thread safety

### ISSUE THREAD

**Issue by CHILI-GmbH** [zeiss-microscopy/libCZI#39]
the docs states that the lib aims to be thread-safe.

does this extend to reading the data from disk? i.e. is it supposed to be possible to use the same cziReader with an opened stream from multiple threads?

doing so gives `Illegal data detected at offset 528800 -> Invalid SubBlockk-magic` errors.

creating new cziReader objects with new streams in each thread avoids these errors, but incurs a considerable overhead.

**Comment by ptahmose** [comment]
Yes, it is supposed to be possible to have one cziReader-object and concurrently "read" from it. One loose end I am aware of: this requires a stream-implementation which allows concurrent calls to IStream::Read. In current code, only the version for Win32 (CSimpleStreamImplWindows) supports this. The other implementations (which are used in Linux e.g. CSimpleStreamImpl) are obviously not thread-safe (since they use an fseek and a subsequent fread, which obviously not a good thing for concurrent access…). A simple fix would be a critical section here, but maybe there is a portable way which is along the lines of the Win32-implementation - I don't know.

- Did you test on Windows (where I'd think it should work out-of-the-box)?
- Can you test with the suggested critical-section (around the Stream-reader-implementation which is used in your case)?

**Comment by ptahmose** [comment]
I had a quick look, and I'd guess that "pread"/"pread64" should do the job -> http://man7.org/linux/man-pages/man2/pread.2.html , https://stackoverflow.com/questions/19780919/read-write-from-file-descriptor-at-offset . However - this would probably require some configuration fiddling (since pread seems to be Linux-only?).
Yes, this indeed should better be mentioned in the documentation.
